New Code Example : Forward-Forward Algorithm for Image Classification

suvadityamuk opened the pull request:
This PR adds a new Code Example implementation for the Forward-Forward algorithm, as introduced by [Prof. Hinton in his paper](https://www.cs.toronto.edu/~hinton/FFA13.pdf) at NeurIPS 2022.

Some things to note:
- This is implemented completely using Keras and TensorFlow-backend code, but is an almost-direct port of [this PyTorch implementation of the same algorithm here](https://github.com/mohammadpz/pytorch_forward_forward)
- Since the weight updates occur within the layer itself as per the algorithm, some of the code may be working against the original design of the Keras API. But such is the algorithm that looks to defy them altogether. Would appreciate any reviews and thoughts on that as well
- I use `TQDM` and `random` for some cleaner visuals and simpler code. 

Tagging @LukeWood @fchollet for a review. Thank you for your time!

Signed-off-by: Suvaditya Mukherjee <suvadityamuk@gmail.com>

suvadityamuk commented:
Hi, @fchollet! Gentle ping for a review, thank you! Oh, and Merry Christmas :christmas_tree: !

fchollet commented:
Thanks for the PR! I'll take a close look in the next few days.

fchollet commented on examples/vision/forwardforward.py:
```diff
@@ -0,0 +1,352 @@
+training instead of the traditionally-used method of backpropagation, as proposed by
+[Prof. Geoffrey Hinton](https://www.cs.toronto.edu/~hinton/FFA13.pdf)
+The concept was inspired by the understanding behind [Boltzmann
+Machines](http://www.cs.toronto.edu/~fritz/absps/dbm.pdf). Backpropagation involves
```
Please keep all markdown links on a single line, otherwise they won't properly render

suvadityamuk replied:
Will make this change across the file

fchollet commented on examples/vision/forwardforward.py:
```diff
@@ -0,0 +1,352 @@
+[Prof. Geoffrey Hinton](https://www.cs.toronto.edu/~hinton/FFA13.pdf)
+The concept was inspired by the understanding behind [Boltzmann
+Machines](http://www.cs.toronto.edu/~fritz/absps/dbm.pdf). Backpropagation involves
+calculating loss via a cost function and propagating the error across the network. On the
```
I think you can find a slightly better one-line description of backprop

suvadityamuk replied:
Sure, will update that.

fchollet commented on examples/vision/forwardforward.py:
```diff
@@ -0,0 +1,352 @@
+
+The following example explores how to use the Forward-Forward algorithm to perform
+training instead of the traditionally-used method of backpropagation, as proposed by
+[Prof. Geoffrey Hinton](https://www.cs.toronto.edu/~hinton/FFA13.pdf)
```
`"Proposed by Hinton in [The Forward-Forward Algorithm: Some Preliminary Investigations]() (2022)"`

suvadityamuk replied:
Right, will make this change

fchollet commented on examples/vision/forwardforward.py:
```diff
@@ -0,0 +1,352 @@
+other hand, the FF Algorithm suggests the analogy of neurons which get "excited" based on
+looking at a certain recognized combination of an image and its correct corresponding
+label.
+This method takes certain inspiration from the biological learning process that occurs in
```
Please add line breaks between paragraphs

fchollet commented on examples/vision/forwardforward.py:
```diff
@@ -0,0 +1,352 @@
+Date created: 2022/12/21
+Last modified: 2022/12/23
+Description: Training a Dense-layer based model using the Forward-Forward algorithm.
+
```
Add an `Accelerator:` field (either GPU or None)

fchollet commented on examples/vision/forwardforward.py:
```diff
@@ -0,0 +1,352 @@
+Forward-Forward algorithm within the layers themselves, instead of the traditional method
+of relying on the global loss functions and optimizers.
+The process is as follows:
+- Perform necessary imports
```
Add line breaks before any list

fchollet commented on examples/vision/forwardforward.py:
```diff
@@ -0,0 +1,352 @@
+"""
+
+
+class FFDense(tf.keras.layers.Layer):
```
You've imported `keras`, so you can replace all instances of `tf.keras` with just `keras`

fchollet commented on examples/vision/forwardforward.py:
```diff
@@ -0,0 +1,352 @@
+        bias_regularizer=None,
+        **kwargs,
+    ):
+        super(FFDense, self).__init__()
```
1. Add `**kwargs` here
2. Use Python 3 style super calls, `super().__init__(**kwargs)`

suvadityamuk replied:
Will do

fchollet commented on examples/vision/forwardforward.py:
```diff
@@ -0,0 +1,352 @@
+            bias_initializer=bias_initializer,
+            kernel_regularizer=kernel_regularizer,
+            bias_regularizer=bias_regularizer,
+            **kwargs,
```
Remove `**kwargs` here

suvadityamuk replied:
Okay, will remove that

fchollet commented on examples/vision/forwardforward.py:
```diff
@@ -0,0 +1,352 @@
+        res = self.dense(x_dir)
+        return self.relu(res)
+
+    def forwardforward(self, x_pos, x_neg):
```
`forward_forward`

suvadityamuk replied:
Right, understood

fchollet commented on examples/vision/forwardforward.py:
```diff
@@ -0,0 +1,352 @@
+        self.num_epochs = num_epochs
+
+    def call(self, x):
+        x_norm = tf.norm(x, ord=2, axis=1, keepdims=True)
```
This assumes 2D data. Could you extend it to data of any dimensionality?

suvadityamuk replied:
Allow me some time to understand how to do so. Since the example is of Image Classification, I took the liberty to keep it that way.

fchollet commented on examples/vision/forwardforward.py:
It's a bit weird to use a layer for this, you can just use a `tf.reshape([tf.shape(h), -1])` op

fchollet commented on examples/vision/forwardforward.py:
Please don't go to numpy for this. For assignment you can use one of the solutions outlined in this thread: https://github.com/tensorflow/community/pull/433

suvadityamuk replied:
Ahh, this looks good, will try to incorporate this

suvadityamuk replied:
Would the use of the XLA op as you have mentioned in the issue be fine?

fchollet commented on examples/vision/forwardforward.py:
```diff
@@ -0,0 +1,352 @@
+                    )
+                )
+                mean_loss = tf.math.reduce_mean(loss)
+                loss_list.append(mean_loss.numpy())
```
no numpy

suvadityamuk replied:
Alright, will replace

fchollet commented on examples/vision/forwardforward.py:
no numpy

fchollet commented on examples/vision/forwardforward.py:
`train_step` should return average values, e.g. the output of `loss_tracker.result()` (where `loss_tracker` is a `metrics.Mean` instance).

Also, use snake case

suvadityamuk replied:
Will make the change

fchollet commented on examples/vision/forwardforward.py:
Do not include any print statements or matplotlib plots in `train_step`. You should write a callback for these.

suvadityamuk replied:
Right. I'll implement a callback and make use of that

fchollet commented on examples/vision/forwardforward.py:
You should arrive to a model that can be run in graph mode, without requiring `run_eagerly=True`

suvadityamuk replied:
Alright, I will try to do so

fchollet commented on examples/vision/forwardforward.py:
What do you get?

suvadityamuk replied:
The Accuracy score currently stood at ~24-27% on different runs (can set a constant Seed to enhance reproducibility)

fchollet replied:
25% accuracy on MNIST tells you that your algorithm does not work, unfortunately. A simple logistic regression does ~92% or so. A simple logistic regression after applying ~90% noise on the input (setting to 0 90% of the pixels, randomly) still does 67%. Even if it were poorly tuned, the example should achieve at least 97% if the algorithm did work as expected.

suvadityamuk replied:
Right. Let me iterate on this algorithm for a while, see if I am making some errors and I will get back to you on this

fchollet reviewed (commented):
Thanks for the PR. It's a great example!

DavidLandup0 commented on examples/vision/forwardforward.py:
```diff
@@ -0,0 +1,352 @@
+    def forwardforward(self, x_pos, x_neg):
+        loss_list = []
+        for i in trange(self.num_epochs):
+            with tf.GradientTape() as tape:
```
Aren't we falling back to backprop here?
The loss is calculated, gradients are computed backwards, and then we're doing an optimizer step

suvadityamuk replied:
This is happening locally, solely for this layer. So wouldn't qualify for _backward_ propagation just yet.

suvadityamuk commented:
Hi, @fchollet 

I have just added a new commit with all the changes. Apologies for the delay, took some time to make sure all of your comments are addressed. 
The new Test Accuracy is 97.75% as per my last training run. I have also made use of the `xla.dynamic_update_slice` as per your linked PR's suggestion (thanks a ton for that, by the way.)
I have also added more detailed comments and suggestions
Please let me know if any other changes are required.

Also, a very Happy New Year to you!

suvadityamuk commented:
Also, not sure why the `docker-image` CI is failing. It seems due to some `keras_cv`-related version issue although I have not used it within this example.

fchollet commented:
> The new Test Accuracy is 97.75% as per my last training run. I have also made use of the xla.dynamic_update_slice as per your linked PR's suggestion (thanks a ton for that, by the way.)

Congrats on making it work!

The current code looks good to me. Please add the generated files. I pushed some copyedits, please pull them first.

suvadityamuk commented:
Hi, have added the generated files and some minor factual edits. Thank you!

fchollet reviewed (approved):
Thank you for the great contribution! 👍

suvadityamuk commented:
Happy to contribute!

michaelStettler commented:
HI, 

First of all thanks a lot for the code. It's very instructive to see how you implemented this directly in keras/tensorflow. Anyway, I wanted to try it and I just copy pasted the code but I got a: 

 KeyError: 'The optimizer cannot recognize variable dense_1/kernel:0. This usually means you are trying to call the optimizer to update different parts of the model separately. Please call `optimizer.build(variables)` with the full list of trainable variables before the training loop or use legacy optimizer 
`tf.keras.optimizers.legacy.{self.__class__.__name__}.'

I've tried several thing from yesterday but none worked. Any suggestion on what that could mean on where to add this optimizer.build?
